lib/ansible/modules/cloud/vmware/vmware_guest_network.py

```diff
@@ -0,0 +1,491 @@
+
+
+class PyVmomiHelper(PyVmomi):
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+        self.change_detected = False
+        self.config_spec = vim.vm.ConfigSpec()
+        self.config_spec.deviceChange = []
+        self.nic_device_type = dict(
+            pcnet32=vim.vm.device.VirtualPCNet32,
+            vmxnet2=vim.vm.device.VirtualVmxnet2,
+            vmxnet3=vim.vm.device.VirtualVmxnet3,
+            e1000=vim.vm.device.VirtualE1000,
+            e1000e=vim.vm.device.VirtualE1000e,
+            sriov=vim.vm.device.VirtualSriovEthernetCard,
+        )
+
+    def get_device_type(self, device_type=None):
+        """ Get network adapter device type """
+        nic_device_type = dict(
+            pcnet32=vim.vm.device.VirtualPCNet32(),
+            vmxnet2=vim.vm.device.VirtualVmxnet2(),
+            vmxnet3=vim.vm.device.VirtualVmxnet3(),
+            e1000=vim.vm.device.VirtualE1000(),
+            e1000e=vim.vm.device.VirtualE1000e(),
+            sriov=vim.vm.device.VirtualSriovEthernetCard(),
+        )
+        if device_type and device_type in nic_device_type:
+            return nic_device_type[device_type]
+        else:
+            self.module.fail_json(msg='Invalid network device_type %s' % device_type)
+
+    @staticmethod
+    def get_parent_datacenter(obj):
+        """ Walk the parent tree to find the object's datacenter """
+        datacenter = None
+        while True:
+            if not hasattr(obj, 'parent'):
+                break
+            obj = obj.parent
+            if isinstance(obj, vim.Datacenter):
+                datacenter = obj
+                break
+        return datacenter
+
+    def network_exists_by_name(self, content, name):
+        """ Check if network with specified name exists """
+        vimtype = [vim.Network]
+        result = find_obj(content, vimtype, name)
+        if result:
+            if to_text(self.get_parent_datacenter(result).name) != to_text(self.params['datacenter']):
+                objects = get_all_objs(content, vimtype)
+                for obj in objects:
+                    if to_text(self.get_parent_datacenter(obj).name) == to_text(self.params['datacenter']):
+                        if to_text(obj.name) == to_text(name):
+                            return True
+            else:
+                return True
+        return False
+
+    def get_network_device_by_mac(self, vm=None, mac=None):
+        """ Get network adapter with the specified mac address"""
+        nic_device = None
+        if vm is None or mac is None:
+            return nic_device
+        for device in vm.config.hardware.device:
+            for device_type in self.nic_device_type.keys():
+                if isinstance(device, self.nic_device_type[device_type]):
+                    if device.macAddress == mac:
+                        nic_device = device
+                        break
```

Comment:
Can it be simpler? E.g. without nic_device/device_type
```python
for device in vm.config.hardware.device:
    if device.macAddress == mac:
        return device
```

Reply:
Yeah, thanks. I think need to find the network device first then checking if it's the desired one. Not all devices have this attribute "macAddress".

Reply:
Maybe wrapping mac comparison with try/except/pass will work on nonexistent attribute?
Just to avoid additional iteration, though I still not fully understand how often are there cases when device with certain device type won't have macAddress attribute.

Reply:
All virtual hardware devices are in this list "vm.config.hardware.device", only network adapter has "macAddress" attribute, so I get network device first, then compare mac address value.

Reply:
I agree that your current approach is definitely better than my suggested. 
Will checking for isinstance with `VirtualEthernetCard` be even better here? 
I've found some of the pyvmomi python code which works like this.
https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.vm.device.VirtualEthernetCard.html

Reply:
Yeah, right, all the types extend this "vim.vm.device.VirtualEthernetCard", I can use it here. Thanks.